pkg/cache/internal/transformers.go

```diff
@@ -0,0 +1,72 @@
+package internal
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+)
+
+type runtimeObjectKind int
+
+const (
+	// prefixed names since there are packages with these names that conflict
+	runtimeObjectKindStructured runtimeObjectKind = iota
+	runtimeObjectKindUnstructured
+	runtimeObjectKindMetadata
+)
+
+func runtimeObjectKindForObject(obj runtime.Object) runtimeObjectKind {
+	switch obj.(type) {
+	case *unstructured.Unstructured:
+		return runtimeObjectKindUnstructured
+	case *unstructured.UnstructuredList:
+		return runtimeObjectKindUnstructured
+	case *metav1.PartialObjectMetadata:
+		return runtimeObjectKindMetadata
+	case *metav1.PartialObjectMetadataList:
+		return runtimeObjectKindMetadata
+	default:
+		return runtimeObjectKindStructured
+	}
+}
+
+type TransformFuncByObject map[runtimeObjectKind]map[schema.GroupVersionKind]cache.TransformFunc
```

Comment:
This type and the enums may be a little confusing. How about this:

```golang
type TransformFuncByObject struct {
    structured TransformFuncByGVK
    unstructured TransformFuncByGVK
    metadata TransformFuncByGVK
}
```

And in the `Set(...)` you can use switch to choose which one should be added in.